R/Spatial.R

```diff
@@ -202,4 +202,156 @@ RLE_Plot <- function(seuratObj, assay = "RNA", sampleIdentifier = "SegmentDispla
+#' @param substructureMetaDataFieldName An annotation that will be concatenated during the results. "Local" FOV information will be concatenated using "fov + substructureMetaDataFieldName + a substructure index" within the columns of the metadata. 
+#' @param summarizeLocalResults An optional boolean that will wrap up the various substructureMetaDataFieldName columns into two single columns. One, which determines if a cell is within ANY of the defined substructures, stored in the output column "Within_Local + substructureMetaDataFieldName". The second is a metadata column that displays which of the local substructures the cell belongs in, concatenated as "Local + substructureMetaDataFieldName". "Local + substructureMetaDataFieldName + 0" is always the noise designation. 
+#' @return Returns a dataframe containing columns related to the substructures found within the images at varying scopes. With summarizeLocalResults = FALSE, (number of FOVs) x (number of subtructures + 1) columns will be added. summarizeLocalResults rolls these high resolution results into two additional columns relative to the fovField. 
+#' @examples
+#' \dontrun{
+#' #Perform Cell Structure detection for B cell follicles. 
+#' 
+#' metadata <- DetectCellStructuresBasedOnCellType(seuratObjectMetadata, 
+#' cellTypeField = "cell_type", 
+#' minimumClusterSizeCoefficient = 0.05,
+#' fovField = "fov",
+#' fovWhitelist = 1,
+#' cellTypeWhiteList = c("Bcell", "B_cell", "B.cell"),
+#' xCoordinateField = "x_FOV_px", 
+#' yCoordinateField = "y_FOV_px", 
+#' substructureMetaDataFieldName = "BCF",
+#' summarizeLocalResults = TRUE
+#' )
+#' 
+#' #load/install packages for plotting
+#' library(pacman)
+#' p_load(ggplot2, dplyr, egg, patchwork)
+#' 
+#' #define plotting layout
+#' layout <- "
+#' #AAAA#
+#' BBBCCC
+#' "
+#' 
+#' #Plot results
+#' ggplot(metadata %>% filter(fov == 1), 
+#' aes(x = x_FOV_px, y = y_FOV_px, color = simple_cellType)) + 
+#' geom_point() + 
+#' egg::theme_article() + 
+#' ggtitle('Cell Type assignment')
+#' ggplot(metadata %>% filter(fov == 1), aes(x = x_FOV_px, y = y_FOV_px, color = factor(Local_BCF))) + 
+#' geom_point() + 
+#' egg::theme_article() +  
+#' ggtitle('Specific substructure cell assignment') + 
+#' ggplot(metadata %>% filter(fov == 1), 
+#' aes(x = x_FOV_px, y = y_FOV_px, color = Within_Local_BCF)) + 
+#' geom_point() + 
+#' egg::theme_article() + 
+#' ggtitle('Non-specific substructure cell assignment') + 
+#' plot_layout(design = layout, guides = "collect")
+#' }
+#' @export
+DetectCellStructuresBasedOnCellType <- function(seuratObjectMetadata, 
+                                                cellTypeField = "cell_type", 
+                                                minimumClusterSizeCoefficient = 0.05,
+                                                fovField = "fov",
+                                                fovWhitelist = NULL,
+                                                cellTypeWhiteList = c("Bcell", "B_cell", "B.cell"),
```

Comment:
I understand B cells drive your use case, but I would not build these kind of assumptions into formal functions. If you need to require a whitelist argument, then make it mandatory to fill out. If you dont need it, make it NULL, and only use it if non-null.

Same thing for substructureMetadataFieldName

Reply:
fair, thanks. I made "Substructure" the default label for identified cell clusters. 

by default the whitelist will now pull all cell types in the metadata field. I suspect the clustering will be pretty poor in general, but it will at least run & the user can refine from there.